istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -222,21 +225,116 @@ func createRemoteSecretFromTokenAndServer(tokenSecret *v1.Secret, clusterName, s
 	secretNamespace := serviceAccount.Secrets[0].Namespace
 	if secretNamespace == "" {
-		secretNamespace = saNamespace
+		secretNamespace = opt.Namespace
+	}
+	return client.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+}
+
+func getOrCreateServiceAccount(client kube.ExtendedClient, opt RemoteSecretOptions) (*v1.ServiceAccount, error) {
+	if sa, err := client.CoreV1().ServiceAccounts(opt.Namespace).Get(
+		context.TODO(), opt.ServiceAccountName, metav1.GetOptions{}); err == nil {
+		return sa, nil
+	} else if !opt.CreateServiceAccount {
+		// User chose not to automatically create the service account.
+		return nil, err
+	}
+
+	if err := createServiceAccount(client, opt); err != nil {
+		return nil, err
+	}
+
+	// Return the newly created service account.
+	sa, err := client.CoreV1().ServiceAccounts(opt.Namespace).Get(
+		context.TODO(), opt.ServiceAccountName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving service account %s.%s after creating it: %v",
+			opt.ServiceAccountName, opt.Namespace, err)
+	}
+	return sa, nil
+}
+
+func createServiceAccount(client kube.ExtendedClient, opt RemoteSecretOptions) error {
+	// Create a renderer for the base installation.
+	r, err := helm.NewHelmRenderer(opt.ManifestsPath, "base", "Base", opt.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed creating Helm renderer: %v", err)
+	}
+
+	if err := r.Run(); err != nil {
+		return fmt.Errorf("failed running Helm renderer: %v", err)
+	}
+
+	values := fmt.Sprintf(`
+global:
+  istioNamespace: %s
+  istiod:
+    enableAnalysis: false
+  configValidation: false
+base:
+  enableCRDTemplates: false
+`, opt.Namespace)
+
+	yamlContent, err := r.RenderManifest(values)
+	if err != nil {
+		return fmt.Errorf("failed rendering base manifest: %v", err)
+	}
+
+	baseFile, err := writeToTempFile(yamlContent)
+	if err != nil {
+		return fmt.Errorf("failed creating manifest file: %v", err)
+	}
+
+	// Before we can apply the yaml, we have to ensure the system namespace exists.
+	if err := createNamespaceIfNotExist(client, opt.Namespace); err != nil {
+		return err
+	}
+
+	// Apply the YAML to the cluster.
+	if err := client.ApplyYAMLFiles(opt.Namespace, baseFile); err != nil {
+		return fmt.Errorf("failed applying manifest %s: %v", baseFile, err)
 	}
-	return kube.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	return nil
+}
+
+func createNamespaceIfNotExist(client kube.Client, ns string) error {
+	if _, err := client.CoreV1().Namespaces().Get(context.TODO(), ns, metav1.GetOptions{}); err != nil {
+		if _, err := client.CoreV1().Namespaces().Create(context.TODO(), &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ns,
+			},
+		}, metav1.CreateOptions{}); err != nil {
+			return fmt.Errorf("failed creating namespace %s: %v", ns, err)
+		}
+	}
+	return nil
+}
+
+func writeToTempFile(content string) (string, error) {
```

Comment:
this could be eliminated with a small restructure of ApplyYAMLFiles - create ApplyYAML which takes string, and ApplyYAMLFiles could take care of reading the file and pass it to ApplyYAML.